docs/internals/optimizer.rst

```diff
@@ -0,0 +1,852 @@
+
+    tag_f:
+      ...body of function f...
+      jump      // out
+
+As long as the body of the function is a continuous basic block, the "Inliner" can replace ``tag_f jump`` by
+the block at ``tag_f`` resulting in:
+
+.. code-block:: text
+
+      tag_return
+      ...body of function f...
+      jump
+    tag_return:
+      ...opcodes after call to f...
+
+    tag_f:
+      ...body of function f...
+      jump      // out
+
+Now ideally, the other optimizer steps described above will result in the return tag push being moved
+towards the remaining jump resulting in:
+
+.. code-block:: text
+
+      ...body of function f...
+      tag_return
+      jump
+    tag_return:
+      ...opcodes after call to f...
+
+    tag_f:
+      ...body of function f...
+      jump      // out
+
+In this situation the "PeepholeOptimizer" will remove the return jump. Ideally, all of this can be done
+for all references to ``tag_f`` leaving it unused, s.t. it can be removed, yielding:
+
+.. code-block:: text
+
+      ...body of function f...
+      ...opcodes after call to f...
+
+So the call to function ``f`` is inlined and the original definition of ``f`` can be removed.
+
+Inlining like this is attempted, whenever a heuristics suggests that inlining is cheaper over the lifetime of a
+contract than not inlining. This heuristics depends on the size of the function body, the
+number of other references to its tag (approximating the number of calls to the function) and
+the expected number of executions of the contract (the global optimizer parameter "runs").
+
+Yul-Based Optimizer Module
+==========================
+
+The Yul-based optimizer consists of several stages and components that all transform
+the AST in a semantically equivalent way. The goal is to end up either with code
+that is shorter or at least only marginally longer but will allow further
+optimization steps.
+
+The optimizer currently follows a purely greedy strategy and does not do any
+backtracking.
+
+All components of the Yul-based optimizer module are explained below.
+The following transformation steps are the main components:
+
+ - SSA Transform
+ - Common Subexpression Eliminator
+ - Expression Simplifier
+ - Redundant Assign Eliminator
+ - Full Function Inliner
+
+Selecting optimizations
+~~~~~~~~~~~~~~~~~~~~~~~
+
+By default the optimizer applies its predefined sequence of optimization steps to
+the generated assembly. You can override this sequence and supply your own using
+the `--yul-optimizations` option:
+
+``` bash
+solc --optimize --ir-optimized --yul-optimizations 'dhfoD[xarrscLMcCTU]uljmul'
+```
+
+Available abbreviations are listed in the `Yul optimizer docs </docs/yul.rst#optimization-step-sequence>`_.
+
+Preprocessing
+~~~~~~~~~~~~~
+
+The preprocessing components perform transformations to get the program
+into a certain normal form that is easier to work with. This normal
+form is kept during the rest of the optimization process.
+
+Disambiguator
```

Comment:
I think this should be "inside" the Preprocessing section, i.e. it should be one header level smaller.

Reply:
adjusted the doc heights throughout the document